torchgeo/datasets/utils.py

```diff
@@ -631,6 +635,74 @@ def path_is_vsi(path: Path) -> bool:
     return '://' in str(path) or str(path).startswith('/vsi')
 
 
+def listdir_vsi_recursive(root: Path) -> list[str]:
+    """Lists all files in Virtual File Systems (VSI) recursively.
+
+    Args:
+        root: directory to list. These must contain the prefix for the VSI
+            e.g. '/vsiaz/' or 'az://' for azure blob storage
+                 '/vsizip/' or 'zip://' for zipped archives
+
+    Returns:
+        A list of all file paths matching filename_glob in the root VSI directory or its
+        subdirectories.
+
+    .. versionadded:: #0.6
+    """
+    dirs = [str(root)]
+    files = []
+    while dirs:
+        dir = dirs.pop()
+        try:
+            subdirs = fiona.listdir(dir)
+            dirs.extend([os.path.join(dir, subdir) for subdir in subdirs])
+        except FionaValueError as e:
+            if 'is not a directory' in str(e):
+                # Assuming dir is a file as it is not a directory
+                # fiona.listdir can throw FionaValueError for only two reasons
+                files.append(dir)
+            else:
+                # fiona.listdir can throw FionaValueError for only two reasons
+                # 1. 'is not a directory'
+                # 2. 'does not exist'
+                raise
```

Comment:
We should be more specific about what we raise here. Is this try-except really the only way to do this? Seems dirty.

Reply:
Would checking the error message and raising our own error be better?

Here is the error that might occur:
https://github.com/Toblerity/Fiona/blob/7490643d145ce0be506a210abd2802fd0fff63f4/fiona/ogrext.pyx#L2171

Reply:
If Fiona accepts [this](https://github.com/Toblerity/Fiona/pull/1436), would this better? (Would still need some workaround until then).

```python
if not fiona.dir_exists(dir):
    continue
elif fiona.isdir(dir):
    subdirs = fiona.listdir(dir)
    dirs.extend([f'{dir}/{subdir}' for subdir in subdirs])
else:
    files.append(dir)
```

Reply:
So the feedback from fiona is that "torchgeo use a dedicated package like fsspec or the Azure Python module." But then we don't get the silver bullet for all of them, and it seems like additional libraries for each file system will have to be installed.

Reply:
We're definitely trying to minimize our dependencies when possible (0.6 actually removes more dependencies than it adds) so I would prefer not to do this. Worst case, users can always use these libraries themselves to get a list of files, then pass these to a GeoDataset (like in 0.5).

Reply:
So what do you think, merge this, or go with my first proposal; make it easier to handle "non-local-files" by adding a method that the users can override?

```python
if os.path.isdir(path):
    pathname = os.path.join(path, '**', self.filename_glob)
    files |= set(glob.iglob(pathname, recursive=True))
elif (os.path.isfile(path) or path_is_vsi(path)) and fnmatch.fnmatch(
    str(path), f'*{self.filename_glob}'
):
    files.add(path)
elif not hasattr(self, 'download'):
    self.handle_non_local_file()
    
def handle_non_local_file(self):
    warnings.warn(
        f"Could not find any relevant files for provided path '{path}'. "
        f'Path was ignored.',
        UserWarning,
    )
```